67/gql filters (#70)
libremfg/libremfg.github.io: content/how-to/gql/mutate.md

```
{{% /tab %}}{{< /tabs >}}

### Upsert

Many `add` operations support _upserting_, which _update_ or _insert_ (create).

> That is, if the object already exists, the operation will update it with the additional fields.

Author: ```suggestion
That is, if the object already exists, the operation will update it with the additional fields.
If the object doesn't exist, the operation will create it.
```

> Besides general UX convenience, upsert is useful when data comes from multiple sources and in no guaranteed order, or when data comes from the message broker.

Author: ```suggestion
Besides general UX convenience, upsert is useful when data comes from multiple sources and in no guaranteed order, like from multiple streams from the message broker.
```

{{% /tab %}}
{{% /tabs %}}

## Deep mutations

You can perform deep mutations at multiple levels.
Deep mutations don't alter linked objects but can add nested new objects or link to existing objects.

For example, this mutation creates a new version of equipment, and associates a new item of equipment with it. Both the `equipmentVersion` and the `equipment` did not exist in the database.

```graphql
  mutation AddEquipmentVersion($addEquipmentVersionInput2: [AddEquipmentVersionInput!]!) {
  addEquipmentVersion(input: $addEquipmentVersionInput2) {
    equipmentVersion {
      id
      equipment {
        id
      }
    }
  }
}
```

**Variables:**

```json
  "addEquipmentVersionInput2": {
    "id": "widget_machine_version_1",
    "version": "1",
    "versionStatus": "DRAFT",
    "equipment": {
      "id": "widget_maker_1",
      "label": "Widget maker 1"

    }
  }
}
```

> You can confirm the record and its nested property exists with a `get` query.

Author: ```suggestion
You can confirm that the record and its nested property exists with a `get` query.
```
